proposals/0279-multiple-trailing-closures.md

```diff
@@ -1,422 +1,369 @@
 ```
 
-The same applies even when using the proposed syntax for a single trailing closure, which provides a trailing syntax for specifying 
-the argument label of a function that accepts a single closure parameter:
+It's important to note that the handling of default arguments in relation to trailing closures is maintained as-is.  For example:
 
 ```swift
-func foo(fn: () -> Void) {
-}
-
-foo {
-  _: { ... }
-}
+func foo(a: () -> Int = { 42 }, b: Any? = nil) {}
 
 foo {
-  fn: { ... }
+  42
 }
 ```
 
-Would be transformed into:
+Although trailing closure matches parameter for `a:` by type, existing trailing closure behavior would match trailing closure argument to parameter labeled as `b:`, which means that previous call to foo is equivalent to:
 
 ```swift
-foo({ ... })
-
-foo(fn: { ... })
+foo(b: { 42 })
 ```
 
-First call is treated as if being a single trailing closure `foo { ... }`, second call uses new syntax to preserve a label, both calls are accepted.
-
-Since all of the essential source information is preserved (locations of all labels and closure blocks) it would be possible for the type-checker to produce diagnostics for invalid code without any changes.
+Now let's add one more parameter to foo to see how this applies to the new multiple trailing closures syntax:
 
-## Source compatibility
-
-This is an additive proposal, which makes ill-formed syntax well-formed but otherwise does not affect existing code.
-
-## Effect on ABI stability
+```swift
+func foo(a: () -> Int = { 42 }, b: Any? = nil, c: () -> Void) {}
 
-This feature is implementable entirely in the parser, as a syntactic transformation on call expressions. It, therefore, has no impact on the ABI.
+foo {
+  42
+} c: {
+  ...
+}
+```
 
-## Effect on API resilience
+Since new type-checking rule dictates a backwards scan starting for the last (labeled) trailing closure before attempting to match an unlabeled argument, this call is equivalent to a following "old" syntax:
 
-Proposed changes do not introduce features that would become a part of a public API.
+```swift
+foo(b: { 42 }, c: { ... })
+```
 
-## Alternatives considered
+This shows that unlabeled trailing closure matching behaves exactly the same way in both scenarios.
 
-It has been noted that code layout conventions can improve the readability of the existing syntax by placing labeled arguments on separate lines e.g.
+There are reasonable arguments against the backwards-scan design for type-checking trailing closures.  Perhaps the strongest argument is that this interaction with default arguments is unintuitive and limiting.  For example, it is natural to want to take a primary, required closure, followed by some optional closures:
 
 ```swift
-when(2 < 3,
-  then: {
-    ...
-    ...
-  },
-  else: {
-    ...
-    ...
-  }
-)
+func resolve(id: UUID,
+             action: (Object) -> (),
+             completion: (() -> ())? = nil,
+             onError: ((Error) -> ())? = nil)
```

Comment:
* Return `-> Void` instead of `-> ()`

```suggestion
func resolve(
  id: UUID,
  action: (Object) -> Void,
  completion: (() -> Void)? = nil,
  onError: ((Error) -> Void)? = nil
) {
  ...
}
```